core/pkg/artifacts/saver.go

```diff
@@ -301,6 +328,119 @@ func newUploadTask(fileInfo serverFileResponse, localPath string) *filetransfer.
 	}
 }
 
+const (
+	S3MinMultiUploadSize = 2 << 30 // 2 GiB, the threshold we've chosen to switch to multipart
+	S3MaxMultiUploadSize = 5 << 40 // 5 TiB, maximum possible object size
+	S3MaxParts           = 10000
+)
+
+func multiPartRequest(path string) ([]gql.UploadPartsInput, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get file size for path %s: %w", path, err)
+	}
+	fileSize := fileInfo.Size()
+
+	if fileSize < S3MinMultiUploadSize {
+		// We don't need to use multipart for small files.
+		return nil, nil
+	}
+	if fileSize > S3MaxMultiUploadSize {
+		return nil, fmt.Errorf("file size exceeds maximum S3 object size: %v", fileSize)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	partsInfo := []gql.UploadPartsInput{}
+	partNumber := int64(0)
+	buffer := make([]byte, getChunkSize(fileSize))
+	for {
+		bytesRead, err := file.Read(buffer)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		if bytesRead == 0 {
+			break
+		}
+		partsInfo = append(partsInfo, gql.UploadPartsInput{
+			PartNumber: partNumber,
+			HexMD5:     utils.ComputeHexMD5(buffer[:bytesRead]),
+		})
+		partNumber++
+	}
+	return partsInfo, nil
+}
+
+func (as *ArtifactSaver) uploadMultipart(path string, fileInfo serverFileResponse,
+	partData []gql.UploadPartsInput, doneChan chan<- *filetransfer.Task,
+) {
+	statInfo, err := os.Stat(path)
+	if err != nil {
+		doneChan <- &filetransfer.Task{Err: err}
```

Comment:
This breaks `Task` conventions. I wouldn't expect `Err` to be set outside the `filetransfer` package, and I also wouldn't expect most fields to be unset.

Since this function always pushes exactly once to `doneChan` and does so synchronously, would it make sense for it to just return a (*Task, error) pair where the Task is nil if there's an error? The pre- and post-conditions on `doneChan` are unclear to me.

Reply:
The way I'm currently doing it I don't need the Task at all, just the name, so I created a separate type to return just the (string, error) pair.